src/lib/pipe_lib/broadcast_pipe.ml

```diff
@@ -0,0 +1,136 @@
+open Core_kernel
+open Async_kernel
+
+type 'a t =
+  { mvar: 'a option Mvar.Read_only.t
+  ; mutable cache: 'a option
+  ; mvar_pipe_handle: 'a option Pipe.Reader.t
+  ; mutable reader_id: int
+  ; pipes:
+      ( 'a option Strict_pipe.Reader.t
+      * ( 'a option
+        , Strict_pipe.drop_head Strict_pipe.buffered
+        , unit )
+        Strict_pipe.Writer.t )
+      Int.Table.t }
+
+let create mvar =
+  let mvar_pipe_handle = Mvar.pipe_when_ready mvar in
+  let t =
+    { mvar
+    ; cache= Mvar.peek mvar |> Option.join
+    ; mvar_pipe_handle
+    ; reader_id= 0
+    ; pipes= Int.Table.create () }
+  in
+  don't_wait_for
+    (Pipe.iter_without_pushback t.mvar_pipe_handle ~f:(fun a ->
+         t.cache <- a ;
+         Int.Table.iter t.pipes ~f:(fun (_, w) -> Strict_pipe.Writer.write w a)
+     )) ;
+  t
+
+exception Already_closed
+
+let guard_already_closed t k =
+  if Pipe.is_closed t.mvar_pipe_handle then raise Already_closed else k ()
+
+let peek t = guard_already_closed t (fun () -> t.cache)
+
+let close t =
+  guard_already_closed t (fun () ->
+      Pipe.close_read t.mvar_pipe_handle ;
+      Int.Table.iter t.pipes ~f:(fun (_, w) -> Strict_pipe.Writer.close w) ;
+      Int.Table.clear t.pipes )
+
+let fresh_reader_id t =
+  t.reader_id <- t.reader_id + 1 ;
+  t.reader_id
+
+let prepare_pipe ~close t =
+  guard_already_closed t (fun () ->
+      let r, w =
+        Strict_pipe.create
+          (Strict_pipe.Buffered (`Capacity 3, `Overflow Strict_pipe.Drop_head))
```

Comment:
I don't think `Drop_head` is what we want here. We'd have to write the consumers of the pipe under the assumption the data is unreliable which is going to severely complicate things.

Reply:
I was thinking consumers don’t really care if they miss a message, but maybe this should be crash instead

Reply:
Changing to synchronous now, so you’ll really not miss a message